### PR TITLE
Heartbeat events

### PR DESCRIPTION
**Pull request by mythmon**
@brittanystoroz This is a fairly old branch. It might have bitrotted, but it *seems* ok to me. What do you think?

r?

### CODE REVIEW

File: .eslintrc.yaml [hunk 1]
```diff
@@ -4,10 +4,10 @@ plugins:
 
 rules:
   indent: [1, 2, {SwitchCase: 1}]
-  mozilla/balanced-listeners: [1]
   mozilla/no-aArgs: [1]
   mozilla/this-top-level-scope: [1]
   no-console: [1]
+  mozilla/balanced-listeners: [0]
```
Comment (mythmon) [inline .eslintrc.yaml]: We'll probably want to not include this, once we get all the cleanup stuff added, like you were working on in #18.

File: lib/Heartbeat.js [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+ *        Whether the heartbeat UI should only be targeted at a private window (if one exists).
+ *        No notifications should be fired when this is true.
+ * @param {String} [options.surveyId]
+ *        An ID for the survey, reflected in the Telemetry ping.
+ * @param {Number} [options.surveyVersion]
+ *        Survey's version number, reflected in the Telemetry ping.
+ * @param {boolean} [options.testing]
+ *        Whether this is a test survey, reflected in the Telemetry ping.
+ */
+exports.Heartbeat = class {
```
Comment (mythmon) [inline lib/Heartbeat.js]: Compare to https://dxr.mozilla.org/mozilla-central/rev/f97a056ae6235de7855fd8aaa04fb1c8d183bd06/browser/components/uitour/UITour.jsm#1046

Comment (brittanystoroz) [inline lib/Heartbeat.js]: Do we need to handle this `privateWindowsOnly` option? https://dxr.mozilla.org/mozilla-central/rev/f97a056ae6235de7855fd8aaa04fb1c8d183bd06/browser/components/uitour/UITour.jsm#1108-1111

File: lib/Heartbeat.js [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+
+    Log.debug('Sending payload to Telemetry: name:', name, 'payload:', payload);
+
+    TelemetryController.submitExternalPing('heartbeat', payload, {
+      addClientId: true,
+      addEnvironment: true,
+    });
+
+    // only for testing
+    this.events.emit('TelemetrySent', payload);
```
Comment (brittanystoroz) [inline lib/Heartbeat.js]: Should we remove this for production? I'm assuming this is just for logging purposes?

Comment (mythmon) [inline lib/Heartbeat.js]: In the original implemention, this was used in automated tests to assert that telemetry was sent as expected. We aren't using it here, but we could. It doesn't produce any logs, so I think it is fine to leave in?

**Comment by brittanystoroz** [comment]
Needs handling of `engagementButtonLabel` option again. I had submitted a PR for this where it simply passed in that option, but UITour was still in control so it all worked fine. Looks like we need to add these two bits of code to our implementation https://dxr.mozilla.org/mozilla-central/source/browser/components/uitour/UITour.jsm#1191-1208 and https://dxr.mozilla.org/mozilla-central/source/browser/components/uitour/UITour.jsm#1289

File: lib/Heartbeat.js (hunk 2 above)
Comment (mythmon) [inline lib/Heartbeat.js]: We aren't using it anywhere yet, so I don't think so.

**Comment by brittanystoroz** [comment]
r+

**Pull request merged**